services/queue/src/api.js

```diff
@@ -300,6 +300,132 @@ builder.declare({
   return res.reply(result);
 });
 
+/** Cancel all tasks in a group */
+builder.declare({
+  method: 'post',
+  route: '/task-group/:taskGroupId/cancel',
+  name: 'cancelTaskGroup',
+  scopes: {
+    AnyOf: [
+      'queue:cancel-task-group:<schedulerId>/<taskGroupId>',
+      'queue:cancel-task-in-project:<projectId>',
+    ],
+  },
+  stability: APIBuilder.stability.experimental,
+  category: 'Tasks',
+  input: undefined,
+  output: 'cancel-task-group-response.yml',
+  title: 'Cancel Task Group',
+  description: [
+    'This method will cancel all tasks with a given `taskGroupId` that are not resolved yet.',
+    'That means all tasks in either `unscheduled`, `pending` or `running` states.',
+    'Behaviour is similar to the `cancelTask` method.',
+    '',
+    '**Remark** this operation does not guarantee that new tasks created for this `taskGroupId`',
+    'would automatically be rejected. For example when running task keeps creating new tasks.',
+  ].join('\n'),
+}, async function(req, res) {
+  const taskGroupId = req.params.taskGroupId;
+  // allow query.reasonResolved ? cancel | superseded
+
+  const [taskGroup] = await this.db.fns.get_task_group(taskGroupId);
+  if (!taskGroup) {
+    return res.reportError('ResourceNotFound',
+      'No task-group with taskGroupId: `{{taskGroupId}}`', {
+        taskGroupId,
+      },
+    );
+  }
+
+  // in order to authorize call we would need to fetch a task to retrieve schedulerId and projectId
+  const someTaskInGroup = Task.fromDbRows(
+    await this.db.fns.get_tasks_by_task_group_projid(taskGroupId, 1, 0),
+  );
+  if (!someTaskInGroup) {
+    return res.reportError('ResourceNotFound',
+      'No tasks found for task-group with taskGroupId: `{{taskGroupId}}`', {
+        taskGroupId,
+      },
+    );
+  }
+
+  await req.authorize({
+    taskGroupId,
+    schedulerId: someTaskInGroup.schedulerId,
+    projectId: someTaskInGroup.projectId,
+  });
+
+  // naïve approach
+  const allTasks = await this.db.fns.get_tasks_by_task_group_projid(taskGroupId, null, null);
+
+  const response = {
+    totalCount: allTasks.length,
+    taskIds: [],
+    cancelledCount: 0,
+    taskGroupId,
+  };
+
+  const allowedStates = ['unscheduled', 'pending', 'running'];
+
+  for (let task of allTasks) {
+    task = Task.fromDb(task);
+    // skip resolved and deadline exceeded tasks
+    if (!allowedStates.includes(task.state()) || task.deadline.getTime() < new Date().getTime()) {
```

Comment:
I don't think you need to check deadline here. If the deadline has passed, but the queue hasn't yet resolved it as deadline-exceeded, I think we would still want to resolve it as cancelled, because it could still be running.

Reply:
true! Thanks